deploy.sh

```diff
@@ -0,0 +1,203 @@
+        exit 1
+    fi
+}
+
+deploy_profile() {
+    echo "=== Deploying profile $PROFILE of node $NODE ==="
+
+    ONLY_PROFILE="del(.path) | del(.activate)"
+    MERGED="$(jq "(. | del(.nodes) | del(.hostname) | $ONLY_PROFILE) + (.nodes.\"$NODE\" | del(.profiles) | $ONLY_PROFILE) + (.nodes.\"$NODE\".profiles.\"$PROFILE\" | del(.hostname))" <<< "$JSON")"
+    HOST="$(get hostname <<< "$MERGED")"
+    SSH_USER="$(get sshUser <<< "$MERGED")"
+    PROFILE_USER="$(get user <<< "$MERGED")"
+    CLOSURE="$(get path <<< "$MERGED")"
+    ACTIVATE="$(get activate <<< "$MERGED")"
+    FAST_CONNECTION="$(get fastConnection <<< "$MERGED")"
+    EXTRA_SSH_OPTS="$(get sshOpts <<< "$MERGED")"
+    BOOTSTRAP="$(get bootstrap <<< "$MERGED")"
+    AUTO_ROLLBACK="$(get autoRollback <<< "$MERGED")"
+
+    ensure_set HOST hostname
+    ensure_set CLOSURE path
+
+    if [[ "$EXTRA_SSH_OPTS" == null ]]; then
+        EXTRA_SSH_OPTS=""
+    fi
+
+    export NIX_SSHOPTS="${EXTRA_SSH_OPTS}"
+
+    SUDO=""
+
+
+    if [[ "$SSH_USER" == null ]]; then
+        if [[ "$PROFILE_USER" == null ]]; then
+            echo "neither user nor sshUser set for profile $PROFILE of node $NODE"
+            exit 1
+        fi
+        SSH_USER="$USER"
+
+    fi
+    if [[ ! "$PROFILE_USER" == null ]] && [[ ! "$PROFILE_USER" == "$SSH_USER" ]]; then
+        SUDO="sudo -u $PROFILE_USER"
+    fi
+
+    if [[ "$PROFILE_USER" == null ]]; then
+        PROFILE_USER="$SSH_USER"
+    fi
+
+
+    if [[ "$PROFILE_USER" == root ]]; then
+        PROFILE_PATH="/nix/var/nix/profiles/$PROFILE"
+    else
+        PROFILE_PATH="/nix/var/nix/profiles/per-user/$PROFILE_USER/$PROFILE"
+    fi
+
+    if [[ "$FLAKE_SUPPORT" == 1 ]]; then
+        nix build --no-link "$REPO#deploy.nodes.$NODE.profiles.$PROFILE.path"
+    else
+        nix-build "$REPO" -A "deploy.nodes.$NODE.profiles.$PROFILE.path" --no-out-link
+    fi
+
+    if [[ -n ${LOCAL_KEY:-} ]]; then
+        nix sign-paths -r -k "$LOCAL_KEY" "$CLOSURE"
+    fi
+
+    if [[ "$ACTIVATE" == null ]]; then
+        ACTIVATE=true
+    fi
+
+    EXTRA_NIX_COPY_OPTS=""
+
+    if [[ ! "$FAST_CONNECTION" == true ]]; then
+        EXTRA_NIX_COPY_OPTS="$EXTRA_NIX_COPY_OPTS --substitute-on-destination"
+    fi
+
+    set -x
+
+    # shellcheck disable=SC2086
+    nix copy $EXTRA_NIX_COPY_OPTS --no-check-sigs --to "ssh://$SSH_USER@$HOST" "$CLOSURE"
+
+    # shellcheck disable=SC2029
+    # shellcheck disable=SC2087
+    # shellcheck disable=SC2086
+    ssh $NIX_SSHOPTS "$SSH_USER@$HOST" <<EOF
+set -euo pipefail
+export PROFILE="$PROFILE_PATH"
+if [[ ! -e "$PROFILE_PATH" ]] && [[ -n "$BOOTSTRAP" ]]; then
+    echo "Bootstrapping"
+    DO_BOOTSTRAP=1
+else
+    DO_BOOTSTRAP=0
+fi
+$SUDO nix-env -p "$PROFILE_PATH" --set "$CLOSURE"
+if [[ "\$DO_BOOTSTRAP" -eq 1 ]]; then
+   eval "set -x; $SUDO $BOOTSTRAP; set +x"
+fi
+set -x
+eval "$SUDO $ACTIVATE" || {
+   if [[ "$AUTO_ROLLBACK" == true ]]; then
+      $SUDO nix-env -p "$PROFILE_PATH" --rollback
```

Comment:
`nix-env --rollback` does not actually work correctly (if you use it more than once), because it reverts to the highest numbered generation, not to the previous one.

For example, there are generations `ment-8` and `ment-9`, and `ment-8` is the active one, because `ment-9` was broken and you reverted it. You create a new generation `ment-10` and activate it, but then decide to rollback again. `nix-env` will rollback to `ment-9` instead of `ment-8`.

Reply:
I guess we can remove the latest generation if it fails